tensorflow/core/ops/image_ops.cc

```diff
@@ -787,6 +787,62 @@ REGISTER_OP("CropAndResizeGradBoxes")
     });
 
 // --------------------------------------------------------------------------
+REGISTER_OP("ROIAlign")
+    .Input("input: float")
+    .Input("rois: float")
+    .Output("output: float")
+    .Attr("spatial_scale: float = 1.0")
```

Comment:
I think most of these attrs should be inputs

Reply:
Why do you think they should be inputs? These are model parameters that doesn't depend on data. Making them inputs would complicate the graph (adding bunch of constant scalar nodes) and would require parsing the same value from the input tensors every execution. Would making them inputs provide any benefits?

Reply:
Anything that does not affect the number or dtype of outputs doesn't need to be an attribute.

Reply:
Thanks, I can change these to inputs. I know it is a hassle but could you please explain the logic behind that? To me it seems quite counter-intuitive to require everything to be input, which in a sense making them dynamic while these arguments should be model hyper-parameters and thus static. Having all model hyper-parameters as ops in the graph seems to be dangerous and hard to protect against since if these attributes become inputs, they can be fed by tensors which change value every batch and thus breaking the model silently. I would really appreciate if you can point me to some documentation of making all such parameters inputs or explain the reasoning behind that. I want to understand it so that I can use it in the future.

Thanks

Reply:
The logic is that attributes should only be used for things on which the number or dtypes of outputs depends. Anything which can be dynamic in a graph should be encodable dynamically.

In this case there's 0 performance overhead to making these dynamic, but we open the possibility that in the future someone might programatically explore this parameter space without recreating graphs.

Reply:
Ok. I will make them inputs. There will be a small overhead but probably negligible. But if these parameters are changed during the training, it will break the training. Somehow the arguments should be ensured that they are fixed for any given model throughout all training. What would be the  way to ensure that?

Reply:
Another question, how can I define defaults for the tensor inputs? The default arguments there are defined by the paper and needs to be at these values for majority of use cases. Do I need to implement a python wrapper myself?

Reply:
Done. But I am not sure if this is a good approach since these attributes are not meant to be variable.

Reply:
Thanks. It doesn't matter whether we think they are variable or not; tf is a complex programming environment and I'm sure someone at some point in the future will want to vary these.

Reply:
I understand that. I am one of the front runners of playing with parameters that are not expected to be played with. But for this calculation varying this parameters will change the problem. If you check the paper, these were not really run-time modifiable parameters.